### PR TITLE
Opennebula host module

### PR DESCRIPTION
**Pull request by rvalle**
##### SUMMARY

New module for OpenNebula cloud platform. This module adds physical hosts to an OpenNebula platform. 

The module is based on the new python bindings for OpenNebula pyone.

Announcement: https://forum.opennebula.org/t/new-addon-pyone/5722

This is the first of a series of modules and special care has been taken in creating shared resources: shared documentation fragments, module utilities, base class, etc.

Integration with ansible-test is also provided, the tests can run against a live OpenNebula platform, by providing a configuration file, or against recorded fixtures. Fixtures are low level XMLRPC wire messages and are provided with tests as .json.gz file. 

ansible-test can also be used in fixture recording mode when used against a live OpenNebula platform.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
one_host

##### ANSIBLE VERSION

```
ansible 2.6.0 (opennebula-host-module c3051c1648) last updated 2018/05/12 07:50:11 (GMT +200)
  config file = None
  configured module search path = [u'/home/rafael/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/rafael/Development/privaz.io/python/ansible/lib/ansible
  executable location = /home/rafael/Development/privaz.io/python/ansible/bin/ansible
  python version = 2.7.14 (default, Mar 14 2018, 13:36:31) [GCC 7.3.1 20180303 (Red Hat 7.3.1-5)]
```

### CODE REVIEW

**Comment by rvalle** [comment]
@mattclay I don't really understand why, but ansibot is asking me to document the argument_spec of the other OpenNebula modules.

I think it has not detected correctly my argument_spec, maybe because it is partly done at the base class? 
I think I need to understand better how this check is performed, and see how can refactor and integrate better.

File: lib/ansible/module_utils/opennebula.py
Comment (resmo) [inline lib/ansible/module_utils/opennebula.py]: Code in module_utils must be bsd licensed.

Comment (rvalle) [reply lib/ansible/module_utils/opennebula.py]: OK, have changed to BSD.

File: lib/ansible/utils/module_docs_fragments/opennebula.py
Comment (kustodian) [inline lib/ansible/utils/module_docs_fragments/opennebula.py]: `one_vm`, `one_image` and `one_image_facts` use the following module parameters/ENV variableis for authentication with the management server:
* `api_url`/`ONE_URL`
* `api_username`/`ONE_USERNAME`
* `api_password`/`ONE_PASSWORD`

We should make sure that all OpenNebula modules which use XMLRPC use the same parameter names and environment variables for authentication.

File: lib/ansible/utils/module_docs_fragments/opennebula.py
Comment (kustodian) [inline lib/ansible/utils/module_docs_fragments/opennebula.py]: Maybe it's better to call this variable similar to other vars e.g. `ONE_HTTPSVERIFY`. Also I don't see a point in having this variable only be used if it's set or not, especially if the default is to check certs. Why not make it override the parameter to whatever value the variable is set true/1, false/0, etc?

File: lib/ansible/modules/cloud/opennebula/one_host.py
Comment (kustodian) [inline lib/ansible/modules/cloud/opennebula/one_host.py]: It should return the host ID if the API returns that and similar information. What is available?

File: test/integration/targets/one_host/tasks/main.yml
Comment (kustodian) [inline test/integration/targets/one_host/tasks/main.yml:89]: If the host doesn't exist and we try to delete it, the module should return `changed: false`.

File: test/integration/targets/one_host/tasks/main.yml
Comment (kustodian) [inline test/integration/targets/one_host/tasks/main.yml]: Maybe test all states for a missing host? It should be easy to test it with a loop.

File: test/integration/targets/one_host/tasks/main.yml
Comment (kustodian) [inline test/integration/targets/one_host/tasks/main.yml:145]: Looking at this, maybe the module could return the template of the host?

File: test/integration/targets/one_host/tasks/main.yml
Comment (kustodian) [inline test/integration/targets/one_host/tasks/main.yml:1]: Maybe you could add a few tests for check mode (`check_mode: yes`) so that we know that the module doesn't change anything in check mode. I remember some Ansible modules having bugs and changing things in check mode.

File: test/integration/targets/setup_opennebula/vars/main.yml
Comment (kustodian) [inline test/integration/targets/setup_opennebula/vars/main.yml:6]: Why do you test with two hosts, isn't one enough?

**Review by kustodian — CHANGES_REQUESTED** [review]
Good stuff. Awesome job with the recorded fixtures, this should make testing of these modules a lot easier even without running OpenNebula.

File: test/integration/targets/setup_opennebula/vars/main.yml
Comment (rvalle) [reply test/integration/targets/setup_opennebula/vars/main.yml:6]: yes, one is enough. It works if you define only one too.
You could also test with many if you wanted to do some kind of "stress" test.
the tests will just run on all the hosts that you list there.

Comment (kustodian) [reply test/integration/targets/setup_opennebula/vars/main.yml:6]: Then it's best to set it to one host.

I would also recommend to keep it simple (the Ansible way) and replace this dictionary with a simple list (e.g. `opennebula_test_hosts`).

Comment (rvalle) [reply test/integration/targets/setup_opennebula/vars/main.yml:6]: Thanks for the feedback, but I actually prefer to test with more than 1 host.
I need to illustrate how to organize fixture data when you run test on multiple items: hosts, vms, images, etc. I expect many other tests to be based on this ones.
However, if you want to run your copy against just one host, it works perfectly.

File: lib/ansible/utils/module_docs_fragments/opennebula.py
Comment (rvalle) [reply lib/ansible/utils/module_docs_fragments/opennebula.py]: This is not a variable we control. It depends on the version of the underlying OS.
 the issue was contributed and reported by @marcindulak 

https://github.com/OpenNebula/addon-pyone/issues/18

The way in which you disable SSL checking requires that variables in some platforms, I believe Centos, for the underlaying XMLRPC subsystem.

Pyone is tested against all the platforms supported by OpenNebula, and this part was a particular pain-point.

File: lib/ansible/utils/module_docs_fragments/opennebula.py
Comment (rvalle) [reply lib/ansible/utils/module_docs_fragments/opennebula.py]: Thanks for the feedback. Yes, we really need to agree on these ones for all the modules.

And we should make it a part of the design principles at: https://github.com/privazio/ansible/wiki/OpenNebula-Ansible-Design-Guidelines

I deliberately chose session following the XMLRPC spec where:

```
All methods expect a session string associated to the connected user as the first parameter.
```

Be aware that Username/Password are specific to the "core" authentication backed. OpenNebula supports other authentication mechanisms, that is why the default nomenclature is a bit generic.

I expect that the token mechanism will be used more than username/password, as tokens can be issued, traced and revoked from the console. 

So I think we should go with SESSION.

However, It would be possible to still take the username / password parameters in order to make it more comfortable for those users that use this authentication backend, as session = "username:password".

The ENDPOINT nomenclature comes from one.conf, the configuration file were users will have to find out were the connection endpoint is, however I don't think URL has any problems either. We just have to pick one. 

For environment variables, I think we should prefix with ONE_* just like the modules are.

What do you think?

File: lib/ansible/modules/cloud/opennebula/one_host.py
Comment (rvalle) [reply lib/ansible/modules/cloud/opennebula/one_host.py]: I am not yet returning any values. And I have left an open point about it to make a decision later in the design guidelines.

File: test/integration/targets/one_host/tasks/main.yml
Comment (rvalle) [reply test/integration/targets/one_host/tasks/main.yml:89]: It is slightly different.
The host is in fact created, the module fails because it does not reach the "ENABLED" state.
Normally this will happen because there is a connectivity problem, and the control node cannot reach the host.
If you run the playbook again, the hsot will still not be "enabled" and your play will continue to fail, util the host reaches the desired state "enabled".

Until you fix the connectivity problem, the one_host module will continue to fail.

On the other hand you can always delete a host even if it is in ERROR state.

File: test/integration/targets/one_host/tasks/main.yml
Comment (rvalle) [reply test/integration/targets/one_host/tasks/main.yml]: Thanks, that is a good idea. Looping different states will make testing easier.

even we can do a transition: enabled, disabled, offline, absent.... 

loads of coverage. will look into it.

File: test/integration/targets/one_host/tasks/main.yml
Comment (rvalle) [reply test/integration/targets/one_host/tasks/main.yml:1]: one_host does not support check mode.
I want to add check mode at a later stage.

**Comment by rvalle** [comment]
mmm acording to Shippable tests failed due to connectivity issues during test setup.

File: lib/ansible/utils/module_docs_fragments/opennebula.py
Comment (kustodian) [reply lib/ansible/utils/module_docs_fragments/opennebula.py]: Looking at the documentation the `session` can either be `username:password` or `username:token`, so I would recommend to make it easier on the users of the module and just have username/password separate. Which means have `api_username`/`api_password` (like other modules have already) and just add an alias to this parameter for `api_token` and an additional `ONE_TOKEN` variable. We could easily modify other modules to have those options, until we switch them to `pyone`.

Regarding endpoint, I'm fine with having an alias for that as well: `api_url` and alias `api_endpoint` (or just `endpoint`). The reason why we decided to use the `api_*` prefix for these connections parameters is because that way in the module docs they will be grouped together (and mostly on top) in the list of parameters, so they are easy to see and understand what they mean (even if you never used OpenNebula).

Comment (kustodian) [reply lib/ansible/utils/module_docs_fragments/opennebula.py]: Looking at the documentation the `session` can either be `username:password` or `username:token`, so I would recommend to make it easier on the users of the module and just have username/password separate. Which means have `api_username`/`api_password` (like other modules have already) and just add an alias to `api_password` called `api_token` and an additional `ONE_TOKEN` variable. We could easily modify other modules to have those options, until we switch them to `pyone`.

Regarding endpoint, I'm fine with having an alias for that as well: `api_url` and alias `api_endpoint` (or just `endpoint`). The reason why we decided to use the `api_*` prefix for these connections parameters is because that way in the module docs they will be grouped together (and mostly on top) in the list of parameters, so they are easy to see and understand what they mean (even if you never used OpenNebula).

File: lib/ansible/utils/module_docs_fragments/opennebula.py
Comment (kustodian) [reply lib/ansible/utils/module_docs_fragments/opennebula.py]: I don't understand how you have no control over this variable, when you have it in the code of the module util? You could just change the name there.

If you ask me I would just drop this whole variable and if you want to disable certificate validation, just set the parameter in the task (as you need to set anyway). I don't see any benefit of this variable when I look at it from the standpoint of the Ansible module.

File: test/integration/targets/one_host/tasks/main.yml
Comment (kustodian) [reply test/integration/targets/one_host/tasks/main.yml:89]: So if I create a host with state `present`, in which state is it created, because there is no state `present`? I guess it's created i state `enabled`. The thing is that this should be documented, as well as other states.

Looking at the code all states except `absent` actually create a new host. So it should be documented like that:
* `absent` deletes the host
* `enabled`, `disabled`, `offline` create a host (if doesn't exist) and set the state, or sets the sate on the current host.
* `present` if the host exists, doesn't change the state, if it doesn't exist creates the host and sets the state to `DEFAULT_STATE` (we could probably discuss what should be the default state, because I guess `enabled` is the default, which from my experience is bad, `disabled` is a much better initial state because it allows for testing of the host before allowing the scheduler to deploy new VMs on it).

Did I get this right?

Also when I think about it, `state` doesn't have to have a default if you ask me, because `present` state isn't natural for host state in OpenNebula.  Maybe we can just remove the `present` state and set `state` to be empty by default. That means that if you just want to make some changes to the host, like the `template` part, you only specify the `name` and the `template` parameter. It would make the docs a lot simpler and easier to understand. I've been using OpenNebula for a long time and it was confusing for me.

File: lib/ansible/modules/cloud/opennebula/one_host.py
Comment (kustodian) [inline lib/ansible/modules/cloud/opennebula/one_host.py]: Since we have this `template` param to be able to change labels/attributes,it would probably be nice to later add `labels` and `attributes` parameters to make it easier on the user.

File: test/integration/targets/setup_opennebula/vars/main.yml
Comment (kustodian) [reply test/integration/targets/setup_opennebula/vars/main.yml:6]: I wasn't thinking about my testing, I'm talking about the tests run by Shippable, no need for them to run on two hosts. Let's make them as fast as possible, they are slow anyway.

**Comment by kustodian** [comment]
I'm not sure I understand GitHub reviews. If I leave a comment one some review comments in the `Files changed` tab, those comments are not displayed here in the `Conversation` tab, and vise versa, if I write a comment here, it's not displayed in the `Files changed` tab.

EDIT: And after a few reloads now they are displayed on both....

File: lib/ansible/utils/module_docs_fragments/opennebula.py
Comment (rvalle) [reply lib/ansible/utils/module_docs_fragments/opennebula.py]: @kustodian we don't control that variable, some platforms python implementation disable SSL certificate check using that variable. You can read here:

https://www.python.org/dev/peps/pep-0493/

We don't use it. except for detecting it. If the variable is present that means that the SSL Validation preference has already been configured in python. And attempting to make it programmaticality fails.

Note that this affects to specific platforms only, mainly Centos.

File: test/integration/targets/one_host/tasks/main.yml
Comment (rvalle) [reply test/integration/targets/one_host/tasks/main.yml:89]: I see where you are going and you see the same problem I did.

The present state is satisfied by hosts that are in Enabled, Disabled or Offline states.

the rationale behind it is becase you want to separate "provisioning" from "maintenance" playbooks.

When I provisiong our datacenter I create the objects: hosts, networks, etc. and Ensure that they are created OK.

In maintenance, for example: applying patches: you are going to disable the host, add OS patches, reboot, and re-enable. (grossly speaking).

If re-run provisioning playbook, you dont want it to change the state of the host to any as long as it is present in the platform.

You don't want your provisioning playbooks to chage the state of the hosts that are already in maintenance as long as it is present.

that is why present is satisfied by Enabled, Disabled and Offline.

File: lib/ansible/modules/cloud/opennebula/one_host.py
Comment (rvalle) [reply lib/ansible/modules/cloud/opennebula/one_host.py]: labels is planned. as per design guidelines. 
what is attributes? isnt it the template but in K=V format?

File: test/integration/targets/one_host/tasks/main.yml
Comment (rvalle) [reply test/integration/targets/one_host/tasks/main.yml:89]: With this point I suggest we take some field experience before deciding forward.

Are you planing on running a physical host cluster? that would be awesome.

File: lib/ansible/utils/module_docs_fragments/opennebula.py
Comment (rvalle) [reply lib/ansible/utils/module_docs_fragments/opennebula.py]: ( I think I already replied this, but now I dont see it)

To be honest I did not like api_*
but you do have a point. I think getting all parameters together and on the top actually brings some value.
We can use the aliases. 
And I am concerned about leaving out session (for other core modules), however, we can add it up later based on field experience.
I will pull an update when I get a chance.

File: lib/ansible/utils/module_docs_fragments/opennebula.py
Comment (rvalle) [reply lib/ansible/utils/module_docs_fragments/opennebula.py]: also... there are many tests with Pyone in all the supported platforms, you can have a go to disable SSL in the platforms that use this variable. It is all already setup in travis.

File: test/integration/targets/setup_opennebula/vars/main.yml
Comment (rvalle) [reply test/integration/targets/setup_opennebula/vars/main.yml:6]: from fixitures is very fast, it is all in memory. have you tried it?

File: lib/ansible/modules/cloud/opennebula/one_host.py
Comment (kustodian) [reply lib/ansible/modules/cloud/opennebula/one_host.py]: Yeah, attributes are that. That's how they are displayed in Sunstone  and that's how we called them for one_vm and one_service modules.

It's cool that those two are planned, they don't have to be here from day 1.

File: test/integration/targets/setup_opennebula/vars/main.yml
Comment (kustodian) [reply test/integration/targets/setup_opennebula/vars/main.yml:6]: I haven't yet, but if it's fast, all is good.

File: test/integration/targets/one_host/tasks/main.yml
Comment (kustodian) [reply test/integration/targets/one_host/tasks/main.yml:89]: I understand the provisioning part, didn't think of that use case completely. Then `present` is fine as is, I would just document what present means when the host is created. I can see that in the future I would like to have parameter e.g. `initial_state` which could define the initial state of the host when it's being created. Having that parameter you wouldn't even have to document what will be the initial state, besides saying that the value of `initial_state` is followed.

We are running a physical host cluster of OpenNebula, and this was the next module we were thinking to work one because we need it for host maintenance which you described.

File: lib/ansible/utils/module_docs_fragments/opennebula.py
Comment (kustodian) [reply lib/ansible/utils/module_docs_fragments/opennebula.py]: Then it's fine, didn't understand how it worked.

File: lib/ansible/modules/cloud/opennebula/one_host.py
Comment (ilicmilan) [inline lib/ansible/modules/cloud/opennebula/one_host.py]: typo

File: lib/ansible/utils/module_docs_fragments/opennebula.py
Comment (kustodian) [reply lib/ansible/utils/module_docs_fragments/opennebula.py]: Awesome! While we were working on our modules we looked at many different other cloud modules and we just checked what were the best things those modules had and copied from them, but also tried to fix the things that were bad. I'm also not the biggest fan of `api_*` names of parameters, they don't feel that they are about OpenNebula, but there are two big upsides which are far more important than that feel:
* When you read them you instantly know for what they are used from their names.
* They are grouped together (and on top) so they are easy to see and it's important that they are easy to see because without them, nothing works.

That's why in the end we decided to call them like that. I think we changed the names of those params during development at least 3 times until we settles with these :)

File: lib/ansible/modules/cloud/opennebula/one_host.py
Comment (rvalle) [reply lib/ansible/modules/cloud/opennebula/one_host.py]: ok I see. the only difference between template and attributes is the original format KV or XML

here I call it template because we inherit a pyone features tha converts dictionaries to XML templates.

I think we can just create an alias to the template called attributes and it would be exactly the same, or not?

Comment (ilicmilan) [reply lib/ansible/modules/cloud/opennebula/one_host.py]: Attributes are KV fields in USER_TEMPLATE section.
Labels are special attribute stored as `LABELS` key and a list of comma separated strings which represents labels.
```
TEMPLATE = [
    LABELS = "label1,label2",
    KEY1 = VALUE1,
    KEY2 = VALUE2
]
```

Comment (kustodian) [reply lib/ansible/modules/cloud/opennebula/one_host.py]: If I remember correctly in our modules `attributes` only support `key=value` (check the docs and examples: https://docs.ansible.com/ansible/devel/modules/one_vm_module.html), we don't support any advance structures, compared to your template, because we want the `attributes` to be used just for that. But yeah they are similar.

File: lib/ansible/modules/cloud/opennebula/one_host.py
Comment (ilicmilan) [reply lib/ansible/modules/cloud/opennebula/one_host.py]: It would be nice to return at least a list of instances ids deployed on that host, so we can connect it with `one_vm`

**Comment by rvalle** [comment]
mmm shipable keeps failing for errors like: time-out connecting to AWS, etc unrelated to this commits.

@resmo do you know how what is the procedure in case this happens?

**Comment by resmo** [comment]
@rvalle thanks for pinging. taking care.

**Comment by resmo** [comment]
closing/reopen to trigger shippable build

File: lib/ansible/modules/cloud/opennebula/one_host.py
Comment (rvalle) [reply lib/ansible/modules/cloud/opennebula/one_host.py]: I believe you are after the flush operation. When you ask a to re-schedule all VMs because you are going to put the host for maintenance? is that the use case?
we can discuss in call next week.

**Comment by rvalle** [comment]
@resmo sorry to bother you again.. but shippable continues to misbehave.

**Comment by resmo** [comment]
@rvalle could you try to rebase to latest devel
~~~
# add an upstream remote, if you haven't set it up already
git remote add upstream git@github.com:ansible/ansible.git

# Update devel branch from upstream
git checkout devel
git pull upstream devel

# Rebase working branch and push to github
git checkout opennebula-host-module
git rebase devel
git push -f origin opennebula-host-module
~~~

**Comment by rvalle** [comment]
@resmo I just rebased, my fingers are crossed!

File: lib/ansible/modules/cloud/opennebula/one_host.py
Comment (kustodian) [reply lib/ansible/modules/cloud/opennebula/one_host.py]: `one_host` module shouldn't return IDs of instances running on it, because this module doesn't work with VMs. Returning a list of all VMs on some host should be part of the `one_vm` module, but we will discuss this next week.

**Comment by kustodian** [comment]
shipit

**Comment by rvalle** [comment]
73902 tests passed. that is quite a lot of them.
Our integration tests are also being run inside a docker and pass.
It took a quite for me to find where, but you can find them in: cloud group 4, that is shippable 66 and 67 for python 2.7 and 3.6

@resmo we are ready to merge this, let me know if any further actions are required.

**Comment by ilicmilan** [comment]
shipit

**Review by resmo — APPROVED** [review]
shipit

**Comment by mattclay** [comment]
@rvalle Thank you for contributing this module with tests. It's always good to see tests with contributions, especially for new modules.

However, one thing that caught my attention (and that of several others on the Ansible Core Team) is that the testing capabilities are directly incorporated into the module/module_utils. This is something that we would like to see moved outside of the module code.

We would like to be able to include your contribution in the upcoming Ansible 2.6 release, which will be reaching code freeze for community contributions on May 25th. If you can have a PR merged to separate the module and tests by the freeze date we will be able to keep your contribution for the 2.6 release. Otherwise we'll need to revert the changes and await a follow-up PR to restore them with the tests separated out.

Given that you already have test fixtures in place, I'm guessing the easiest way to approach this would be to convert the tests to unit tests. If you'd like further suggestions on how to approach this or have any other questions, feel free to ask here or in #ansible-devel on IRC.

**Comment by rvalle** [comment]
@mattclay you are welcome.

Yes, I don't like the test traces in production either. I will split the testing code form module utils.

it is mainly "activation" code to tweak the fixture subsystem.

I tried to place it originally in the testing classes: CloudEnvironment, CloudPlatform, etc. But it took me a while to figure out how they work for the basic tasks. 

Let me have a look, and ask around. It should not be complicated.

I have created this issue to handle it:
https://github.com/privazio/ansible/issues/6

**Pull request merged**
